contracts/src/FunctionGatewayV2.sol

```diff
@@ -0,0 +1,294 @@
+    uint32 public nonce;
+
+    /// @dev A mapping from request nonces to request hashes.
+    mapping(uint32 => bytes32) public requests;
+
+    /// @dev The current function identifier.
+    bytes32 public currentVerifiedFunctionId;
+
+    /// @dev The current function input hash.
+    bytes32 public currentVerifiedInputHash;
+
+    /// @dev The current function output.
+    bytes public currentVerifiedOutput;
+
+    /// @dev Initializes the contract.
+    /// @param _timelock The address of the timelock contract.
+    /// @param _guardian The address of the guardian.
+    function initialize(
+        address _timelock,
+        address _guardian
+    ) external initializer {
+        __TimelockedUpgradeable_init(_timelock, _guardian);
+    }
+
+    /// @dev Creates a onchain request for a proof. The output and proof is fulfilled asynchronously
+    ///      by the provided callback.
+    /// @param _functionId The function identifier.
+    /// @param _input The function input.
+    /// @param _context The function context.
+    /// @param _callbackSelector The selector of the callback function.
+    function request(
+        bytes32 _functionId,
+        bytes memory _input,
+        bytes memory _context,
+        bytes4 _callbackSelector
+    ) external payable returns (bytes32) {
+        return
+            request(
+                _functionId,
+                _input,
+                _context,
+                _callbackSelector,
+                DEFAULT_GAS_LIMIT
+            );
+    }
+
+    /// @dev Creates a onchain request for a proof. The output and proof is fulfilled asynchronously
+    ///      by the provided callback.
+    /// @param _functionId The function identifier.
+    /// @param _input The function input.
+    /// @param _context The function context.
+    /// @param _callbackSelector The selector of the callback function.
+    /// @param _callbackGasLimit The gas limit for the callback function.
+    function zkRequest(
+        bytes32 _functionId,
+        bytes memory _input,
+        bytes memory _context,
+        bytes4 _callbackSelector,
+        uint32 _callbackGasLimit
+    ) external payable returns (bytes32) {
+        // Compute the callback hash uniquely associated with this request.
+        bytes32 inputHash = sha256(_input);
+        bytes32 contextHash = sha256(_context);
+        address callbackAddress = msg.sender;
+        bytes32 requestHash = _requestHash(
+            nonce,
+            _functionId,
+            inputHash,
+            contextHash,
+            callbackAddress,
+            _callbackSelector,
+            _callbackGasLimit
+        );
+
+        // Store the callback hash.
+        requests[nonce] = requestHash;
+        emit Request(
+            nonce,
+            _functionId,
+            _input,
+            _context,
+            _callbackSelector,
+            _callbackGasLimit
+        );
+
+        // Increment the nonce.
```

Comment:
I think its better to increment before the saving & emitting the request. It's not a huge deal but (1) never have zero nonce request (2) if caller needs to check nonce again after zkRequest for some reason, it's for their own.

Reply:
resolved.